lib/platform/linux-uart.c

```diff
@@ -0,0 +1,441 @@
+/*
+ * Microsemi Switchtec(tm) PCIe Management Library
+ * Copyright (c) 2017, Microsemi Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef __linux__
+
+#include "../switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "gasops.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stddef.h>
+#include <assert.h>
+#include <string.h>
+
+#include <termios.h>
+#include "../crc8.h"
+
+const char *dbginfo_off = "pscdbg 0 all\r";
+const char *echo_off = "echo 0\r";
+const char *echo_on = "echo 1\r";
+
+struct switchtec_uart{
+	struct switchtec_dev dev;
+	int fd;
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(50)
+
+static int cli_control(struct switchtec_dev *dev, const char *str)
+{
+	int ret;
+	int idx;
+	int rty_cnt = 0;
+	char rtn[1024];
+	struct switchtec_uart *udev = to_switchtec_uart(dev);
+
+	ret = tcflush(udev->fd,TCIOFLUSH);
+	if (ret < 0)
+		return ret;
+
+	ret = write(udev->fd, str, strlen(str));
+	if (ret != strlen(str))
+		return -1;
+
+	ret =  tcdrain(udev->fd);
+	if (ret < 0)
+		return ret;
+
+	memset(rtn, 0, sizeof(rtn));
+
+cli_rtn_retry:
+	if (rty_cnt > RETRY_NUM)
+		return -1;
+
+	ret = read(udev->fd, rtn, sizeof(rtn));
```

Comment:
Rely on specific symbol appeared in the response string.
For example: 
by sscanf(rtn, "%*[^:]:%x>", &idx), check the ":" in the string.
Following are cli commands like echo and debug print off:
 
0x00000000:121c>echo 1
Echo - ON
0x00000000:121d>echo 0
Echo - OFF
0x00000000:121e>
Echo - ON
0x00000000:121f>pscdbg 0 all
Mask before 00000000
Mask after 00000000
0x00000000:1220>

Reply:
I understand that but it doesn't really change things. On a closer look though it looks like you are using an inter-byte time out which shouldn't be as slow as I thought. However, you should definitely put the read in a loop until you see the end of the data you are looking for.

Reply:
Hi, Logan,

The code is put the read in a loop(for gasrd/gaswr/cli_control), until see the end of the data looking for.
Where counter is used to avoid unlimited loop in case of the low probability case: the specific symbol losted for the reason of a bad link.

************************************
preseduo code:

retry the read:
               if (retry counter saturated)
                             exit(1);
               
               return str = read();

               check the if the specific symbol exists in the "return str"
               Yes: go on
               No: loop to "retry the read"

************************************

Reply:
Yes, but what happens if you get the first part of the prompt in one read and the second part of the prompt in another read? The scanf is going to fail both times even though you got the data back you expected.

Reply:
For different cli commands, the output(lines, patterns, symbols) is different.

Reply:
Yes, so the point is you probably want to read line by line until you find a prompt. Not try to fit all possible output into 1024 characters and hope there's a prompt by the end of it....

Reply:
For cli_control, utility does not care the output, but just the end of prompt.
But for gasrd/gaswr, utility firstly care if the end of prompt comes, if so, then abstract the need data from the buffer which collect entire data, such as the crc and data payload for gasrd.
This is why I chose read all data instead of line by line.